tests/tests.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#[macro_use]
+extern crate fail;
+
+use std::sync::mpsc;
+use std::time::*;
+use std::*;
+
+#[test]
+fn test_off() {
+    let f = || {
+        fail_point!("off", |_| 2);
+        0
+    };
+    assert_eq!(f(), 0);
+
+    fail::cfg("tests::off", "off").unwrap();
+    assert_eq!(f(), 0);
+}
+
+#[test]
+fn test_return() {
+    let f = || {
+        fail_point!("return", |s: Option<String>| {
+            s.map_or(2, |s| s.parse().unwrap())
+        });
+        0
+    };
+    assert_eq!(f(), 0);
+
+    fail::cfg("tests::return", "return(1000)").unwrap();
+    assert_eq!(f(), 1000);
+
+    fail::cfg("tests::return", "return").unwrap();
+    assert_eq!(f(), 2);
+}
+
+#[test]
+fn test_sleep() {
+    let f = || {
+        fail_point!("sleep");
+    };
+    let timer = Instant::now();
+    f();
+    assert!(timer.elapsed() < Duration::from_millis(1000));
+
+    let timer = Instant::now();
+    fail::cfg("tests::sleep", "sleep(1000)").unwrap();
+    f();
+    assert!(timer.elapsed() > Duration::from_millis(1000));
```

Comment:
if our test run quickly, maybe we should use `>=` to check here.

Reply:
Actually it's impossible. There are many instructions before sleeping which may consumes 100ns+ time.